### ISSUE TITLE
Tor Integration Not Working - Database Connection

### ISSUE THREAD

**Issue by KitchM** [keepassxreboot/keepassxc#7319]
## Overview
Database integration is not available in Tor

## Steps to Reproduce
1.Click on KeePassXC icon on bar 
2. See error message
3. 

## Expected Behavior
Should have automatically connected to the running KeePassXC instance.

## Actual Behavior
I receive the message;
KeePassXC-Browser has encountered an error:

Cannot connect to KeePassXC. Check that browser integration is enabled in KeePassXC settings.

I select the Reload button:
KeePassXC-Browser has encountered an error:

Key exchange was not successful. 

## Context
KeePassXC - Version 2.6.6
Revision: 9c108b9
Distribution: AppImage

Qt 5.15.2
Debugging mode is disabled.

Operating system: Debian GNU/Linux 11 (bullseye)
CPU architecture: x86_64
Kernel: linux 5.10.0-10-amd64

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- libgcrypt 1.8.1

Operating System: Linux
Desktop Env: XFCE
Windowing System: X11

**Comment by KitchM** [comment]
Note, this procedure worked well in Firefox.  All it did was ask for a name.  In Tor I never get that far.  Reinstalled extension in both, and upgraded the program.

**Comment by KitchM** [comment]
From the Connected Databases section of the settings:
      No connected databases found.
Databases connected to KeePassXC-Browser.
Connect.

Selecting the Connect button does nothing.

**Comment by droidmonkey** [comment]
You likely need to define a custom browser location for tor. Tor installs to many different locations depending on its distribution. Use the custom browser option under the advanced browser settings. You need to specify the native messaging configuration folder. Refer to Tor documentation for that location.

**Comment by varjolintu** [comment]
The [Troubleshooting Guide](https://github.com/keepassxreboot/keepassxc-browser/wiki/Troubleshooting-guide) says:
> Tor Browser installed via the official .tar.gz package will look for the following folders (when extracted to the home folder): /home/<user>/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts/ and globally from /usr/lib/mozilla/native-messaging-hosts/.

> Tor Browser installed via Torbrowser-launcher looks for the following (for en_US): /home/<user>/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts/

> For now, the .json file needed can be symlinked or copied to those directions, for example from under .mozilla/native-messaging-hosts.

**Comment by pentamassiv** [comment]
The Arch AUR package changed its install location and that also broke the connection for me. Today I finally looked into it and fixed it by opening KeepassXC and going to the settings > Browser Integration > Advanced and activating the option "Use a custom browser configuration location" and pasting the following path: `~/.local/opt/tor-browser/app/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts`. Selecting the regular "Tor Browser" does not work and it can be deselected.

I couldn't find documentation on where Tor Browser expects the files to appear so I looked up the install path and searched for the folder `.mozilla` in it. In this folder the browser expects the native-messaging-hosts folder to be in. You should be able to find the correct path with this approach if my path is different from yours.

**Comment by droidmonkey** [comment]
Why can't the Tor project pick a standard and figure this out...

**Comment by KitchM** [comment]
Totally agree.  It does seem obvious, and trivial to fix.

**Comment by KitchM** [comment]
I am forced to reopen this issue because the steps I followed from the directions did not work.

The information from trouble-shooting guide has this part that is applying to my situation on Debian Bullseye:
"Tor Browser installed via Torbrowser-launcher looks for the following (for en_US): /home/<user>/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts/"

My native Firefox has this location:
/home/<user>/.mozilla/native-messaging-hosts/

So I created a symlink like this:
ln -s /home/<user>/.mozilla/native-messaging-hosts/ /home/<user>/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts

I then restarted Tor and could still not get the database seen by the plugin.

What did I miss?

**Comment by droidmonkey** [comment]
https://github.com/keepassxreboot/keepassxc/issues/7319#issuecomment-1064713385

**Comment by KitchM** [comment]
Tried that, but no change.  "No connected databases found."  I haven't a clue about the rest.

BTW,  sudo strace -f -p $(pgrep firefox) 2>&1 | grep keepass
returns nothing.

**Comment by droidmonkey** [comment]
Get rid of your symlink

**Comment by KitchM** [comment]
I removed it and now when I try to click on the icon the message is:
"Cannot connect to KeePassXC. Check that browser integration is enabled in KeePassXC settings."  So I select "Reload" and get:
"Key exchange was not successful."

I'm totally confused.  The guide said to create the symlink.  I wish there were a simple series of steps to follow.

**Comment by droidmonkey** [comment]
It's really easy, you go to the advanced part of the browser integration settings then enter the path to the native messaging folder under the custom browser section. Make sure that is the correct folder for your browser. Also make sure your browser is not running in a restricted sandbox or firejail.

**Comment by KitchM** [comment]
Nothing changes.

Tor 11.5.1 based on Mozilla Firefox 91.12.0 esr.  The OS is Debian 11 (Bullseye) with Xfce 4.16.  The path to the Firefox json file is `/home/user/.mozilla//native-messaging-hosts/`.

**Comment by droidmonkey** [comment]
You don't point to the Firefox file, you point to the Tor native messaging directory. See above for the path, it's non trivial

**Comment by KitchM** [comment]
"Browser Plugin Failure: Could not save the native messaging script file for custom."

I don't understand why KeepassXC needs to know where it is when the troubleshooting guide already does.  Is it not programmed into the app?

**Comment by droidmonkey** [comment]
Tor puts their config directory in all sorts of locations depending on distro and install method.

**Comment by KitchM** [comment]
Yeah, I got that before.  But I wonder if it would solve the problem by just creating a simple list file of these locations which Keepassxc could parse to find it there are any matches?  Wouldn't that solve the problem for Tor?

**Comment by varjolintu** [comment]
> Yeah, I got that before. But I wonder if it would solve the problem by just creating a simple list file of these locations which Keepassxc could parse to find it there are any matches? Wouldn't that solve the problem for Tor?

With Tor this could be quite tricky because they often include the locale in the path.

**Comment by KitchM** [comment]
As far as I can tell, I've tried everything suggested.  Does anyone have any other ideas how to make this work?

**Comment by varjolintu** [comment]
I'd try to double check the Tor path with `strace` with the correct binary name.

**Comment by KitchM** [comment]
What is the best way to determine the correct binary name?

**Comment by varjolintu** [comment]
Start Tor, run `ps -aux` and see what the executable name is? Of course XFCE's system monitor etc. should also show it.

**Comment by KitchM** [comment]
The basic information I was able to gather was:

debian-+    1370  0.0  0.3  62008 54500 ?        Ss   10:35   0:03 /usr/bin/tor --defaults-torrc /usr/share/tor/tor-service-defaults-torrc -f /etc/tor/torrc --RunAsDaemon 0

user       6810  0.0  0.0   6888  3228 ?        S    10:37   0:00 bash /home/user/Programs/tor-browser_en-US/Browser/start-tor-browser

**Comment by KitchM** [comment]
Sorry to be a pain, but I still got nothin'.  Any direction is appreciated.

**Comment by varjolintu** [comment]
So previously you used `/home/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts/` as the script location but `strace` tells it's `/home/user/Programs/tor-browser_en-US/Browser/` ?

**Comment by KitchM** [comment]
I am confused enough that I'm starting over.

So let's see; I need to find the binary name or location of tor.  Tor info shows:
Name 	Firefox
Version 	91.12.0esr
Build ID 	20220607020101
Application Binary 	/home/user/Programs/tor-browser_en-US/Browser/firefox.real
The directory is /home/user/Programs/tor-browser_en-US/Browser/TorBrowser/

**Comment by varjolintu** [comment]
I would assume the correct path then is `/home/user/Programs/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts/`.

**Comment by KitchM** [comment]
I am finding native-messaging-hosts all over my filesystem.  The most relevant ones appear to be:
/home/user/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts

/home/user/.mozilla/native-messaging-hosts

/home/user/~/.local/opt/tor-browser/app/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts

**Comment by varjolintu** [comment]
@KitchM Of course you find them because you probably tried to install them to multiple different locations.

**Comment by KitchM** [comment]
I know for a fact that I never did that.

**Comment by varjolintu** [comment]
@KitchM If you used the KeePassXC's Custom Browser option, that creates those files.

**Comment by KitchM** [comment]
Okay.  The question then becomes if I should have done that or not.  I mean, could the Tor integration work without that.

**Comment by varjolintu** [comment]
It needs the script file to work. You'll just need to figure out the right path.

**Comment by KitchM** [comment]
Ah, so back where we were.

So the issue, as I understand it, is that the basic location of the native-messaging-hosts that Tor creates is not the correct one.  I must find one from Firefox, and preferably from the Firefox that is working with KeepassXC right now.  Do I have the right?

**Comment by varjolintu** [comment]
You'll have to find the one Tor Browser uses. You already got few good hints I suggested earlier based on your `strace` output.

**Comment by KitchM** [comment]
But the one it uses does not connect to the database.  I still don't understand.

**Comment by KitchM** [comment]
Just tried activating integration with Chromium and got the message "Could not save the native messaging script file for custom".

**Comment by chymian** [comment]
hello,
having the same pbls. on debian bullseye, with tor-browser en-US.
the file: `org.keepassxc.keepassxc_browser.json` is created by keepassxc under:
`~/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts`
where it should be, AFAIK.
getting the message:  `Message encryption failed. Is KeePassXC running? `

OS: debian bullseye
keepassxc:  2.6.2+dfsg.1-1, (deb install)
torbrowser-launcher  0.3.3-6 (deb) 
Tor Browser 11.5.1

**Comment by varjolintu** [comment]
@chymian The script file is created by KeePassXC but it still doesn't mean it's the correct path for your Tor Browser installation.

**Comment by chymian** [comment]
@varjolintu 
I was relating to that note from @KitchM :
> I am finding native-messaging-hosts all over my filesystem. The most relevant ones appear to be: /home/user/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts
> 
it's an assumption: since it's the only similiar path within a firefox directory, as in the normal FF-setup::
```
ls -a ~/.mozilla/
.  ..  extensions  firefox  native-messaging-hosts  systemextensionsdev
cd ~/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser
ls -a .mozilla/
.  ..  systemextensionsdev
```
and from the troubleshoting guide:
```
~/.tor-browser/app/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts
```
it's always under `.../Browser/TorBrowser/Data/Browser/.mozilla/`

**Comment by KitchM** [comment]
And still the question remains: exactly how does one rectify the situation?  Where is the definitive step-by-step guide to fix the problem?

**Comment by varjolintu** [comment]
The step-to-step guide is to check the correct path with `strace` and use that with the Custom Browser option. There's nothing more we can do to help. The path depends entirely on your system and the installation path.

**Comment by KitchM** [comment]
Where does one enter the path?

**Comment by varjolintu** [comment]
Settings -> Browser Integration -> Advanced -> [ ] Use custom browser configuration location

**Comment by chymian** [comment]
@varjolintu, @KitchM 
strace showed, that the location is the right one for debian bullseye, en_US (depends on your language):
`/home/user/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/.mozilla/native-messaging-hosts`

it also revealed an access denied, short after

searching here in the issues, showed that this is an app-armor problem, which has already an solution [in issue 3096](https://github.com/keepassxreboot/keepassxc/issues/3096#issuecomment-778816892)
I quote for more easy access, what @StableNarwhal has written: 

one need `apparmor-utils` installed.

> It's been a while, so I won't be able to provide an in-depth explanation without researching again myself, unfortunately.
But what these terminal commands basically do is this:

> 1.    By invoking aa-complain on your firefox instance, you set the enforcement mode to complain mode. In this mode security policy is not enforced but rather access violations are logged to the system log.
> 2.    By invoking aa-logproof, you are able to update your existing AA profile based on logged violations (We turned on logging in step 1). Since keypassxc-proxy will be logged as one of those violations, we are able to modify the profile rules regarding the process. By setting it to (C)hild, we allow it to run it with a sub-profile of the parent process, effectively allowing communication via native messaging only for our proxy.
> 3.    By invoking aa-enforce we turn the strict enforcement of the new profile back on, this time with the added exception.

> I hope this was helpful!

since I don't now a lot, about app-armor, I give no advise on how to exactly deal with all the complained logs.
Basically, I allowed the obfs4proxy (I)nherit & keepassxc stuff as (c)hild. 
maybe one of the devs can give a profound advice here.
